### PR TITLE
rustdoc: clean up source sidebar hide button

### PR DESCRIPTION
**Pull request by notriddle**
This is a redesign of the feature, with parts pulled from https://github.com/rust-lang/rust/pull/119049 but with a button that looks more like a button and matches the one used on other sidebar pages.

Preview:

* http://notriddle.com/rustdoc-html-demo-8/source-sidebar-resize/src/std/lib.rs.html
* http://notriddle.com/rustdoc-html-demo-8/source-sidebar-resize/std/index.html

| | Before | After |
|--|--|--|
| Closed | image | image
| Open | image | image
| Mobile Closed | image | image
| Mobile Open | image | image

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I found the previous design way more obvious. In this case, nothing indicates that:
 1. There is a sidebar
 2. The sidebar is not about the content but about files hierarchy

Also I really don't like the appearance when sidebar is opened for the title and close button (that is very much personal taste).

**Comment by GuillaumeGomez** [comment]
> Is this about the empty vertical column when the sidebar is closed, or is it about the icon?
>
> I had removed the vertical column because it ate into horizontal screen real estate, which is at a premium since code isn't line wrapped. But if you find that helpful, it could be redesigned to put it back.

On mobile the vertical column isn't visible and I think the problem is quite limited on desktop. But yes, I think it'd help to make the connection between the button and the sidebar.

> The `>` icon doesn't seem better about that.

Indeed.

> I'm hoping to lean on people's existing familiarity with GitHub.

And because of this comment, I just realized that this sidebar could be collapsed. I'm not sure if it's bad UI or just me who's very bad at that... :sweat_smile: 

> The fact that this particular icon doesn't seem to have a tooltip in GitHub's design should be criminal, and this PR does include a tooltip for this exact reason. But the icon isn't exactly the same (it's closer to [Apple's](https://developer.apple.com/design/human-interface-guidelines/sidebars)).

The button doesn't stand out much either in their case, hence why I never paid attention to it.

**Comment by GuillaumeGomez** [comment]
> No problem. I've just pushed a new version of the preview docs and PR with the vertical placeholder column.

Much better for when the sidebar is closed. Now my only question is about the mobile version: currently we can open the sidebar wherever we are on the page. With your change, it's not possible anymore. I think to be taken into consideration as some source code pages can be quite huge and if you're at the bottom, it's quite annoying to go back to the top. And also, it makes the sidebar harder to "discover" if you're not already aware of its existence in this case.

Now for when the sidebar is open, I think the title should still be centered and the button remaining on the left side (as currently). You also removed the line which was used as marker to differentiate the "title" and the "content" before. Any reason for that change?

> Do you think this is because:
> 
>     * The buttons doesn't look like buttons? GitHub's buttons are pretty similar to ours, and if they don't look like buttons, then we should probably add more shading or otherwise try to improve clickability signifiers.

I think it would be already a pretty good improvement. I think it's because there is not a strong enough visual marker that I don't notice it.

>     * The icon is bad? What does a good sidebar icon look like? (hamburger buttons should open a modal slide-out, not toggle a persistent navigation panel)

Not a strong opinion on this one. But in this case, I think we should use the same icon for both mobile and desktop, whichever it is.

>     * This particular button is badly designed for some other reason?

Since you moved it back inside a "column" on desktop, looks good to me. On mobile I still have worries I described above.

**Comment by GuillaumeGomez** [comment]
> That's odd. What browser are you testing on? The hamburger button is supposed to stick while you scroll, the same way the `>` button did.

After a force refresh, problem gone. All good there. :)

> I prefer left alignment because the docs sidebar does it the same way.

Depends on the "mode": on mobile, the button is on the left and the title is on the right. There is no equivalent for desktop to compare to unfortunately.

> Not really. If you want it, I'll put it in.

Let's see how it looks with it then. :+1:

**Comment by GuillaumeGomez** [comment]
> Backing up and trying to think of a different approach, maybe we should use an icon based on its purpose instead of its mechanics, which is the same across all viewport sizes.

That's a very good point! I actually like the new icon quite a lot.

File: src/librustdoc/html/static/css/noscript.css [hunk 1]
```diff
@@ -11,7 +11,7 @@ rules.
 
 #copy-path, #sidebar-button, .sidebar-resizer {
 	/* It requires JS to work so no need to display it in this case. */
-	display: none;
+	display: none !important;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/css/noscript.css:14]: Why this change?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Apart from the question I asked, looks good to me, thanks! Once it's answered, we can start the FCP.

File: src/librustdoc/html/static/css/noscript.css (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/css/noscript.css:14]: :+1:

**Comment by GuillaumeGomez** [comment]
Time to start the FCP then.

@rfcbot fcp merge

**Comment by Nemo157** [comment]
> The fact that this particular icon doesn't seem to have a tooltip in GitHub's design should be _criminal_

Interestingly they _do_ have a tooltip on this exact icon in the PR-diff-view.

The biggest thing I notice with this (and github's similar button) is that it feels wrong to me that it's on the left, the button to close a panel should be in the top-right. But that's likely platform-specific behavior, and I don't think rustdoc has any similar UI anywhere to be self-consistent with (unlike github which uses top-right-close everywhere I can see except this one panel).
